src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala

```diff
@@ -12,12 +14,27 @@ trait SparkRowExtractor {
    * @param datum input row
    * @return list of feature keys
    */
-  def getKeyFromRow(datum: GenericRowWithSchema): Seq[String]
+  def getKeyFromRow(datum: Any): Seq[String]
 
   /**
    * Get the feature value from the row
    * @param datum input row
    * @return A map of feature name to feature value
    */
-  def getFeaturesFromRow(datum: GenericRowWithSchema): Map[String, FeatureValue]
+  def getFeaturesFromRow(datum: Any): Map[String, FeatureValue]
+
+  /**
+   * Whether the Row extractor needs a one-time batch preprocessing before calling
+   * getKeyFromRow() and getFeaturesFromRow().
+   * This is especially useful when users need to convert the source DataFrame
+   * into specific datatype, e.g. Avro GenericRecord or SpecificRecord.
+   */
+  def hasBatchPreProcessing() = false
+
+  /**
+   * One time batch preprocess the input data source into a RDD[_] for feature extraction later
+   * @param df input data source
+   * @return batch preprocessed dataframe, as RDD[IndexedRecord]
+   */
+  def batchPreProcess(df: DataFrame) : RDD[IndexedRecord] = throw new NotImplementedError("Batch preprocess is not implemented")
```

Comment:
IMO the naming `batchPreProcess` is too general for what this function does. The naming implies that this function can be used for batch preprocessing of data but it is really supposed to be a conversion function between DataFrame -> RDD[IndexdRecord]. Not a blocker but I think something like `convertDfToRdd` would better reflect the purpose of this function.

Reply:
Changed to convertToAvroRdd